From 8a31bcd1b07f6d541e108623dffc67579798b105 Mon Sep 17 00:00:00 2001
From: Akira TAGOH <akira@tagoh.org>
Date: Thu, 15 Jan 2015 16:42:41 +0900
Subject: [PATCH 21/52] Fix an assertion in GIO

---
 imsettings-daemon/imsettings-server.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/imsettings-daemon/imsettings-server.c b/imsettings-daemon/imsettings-server.c
index 25f45a6..300bbe0 100644
--- a/imsettings-daemon/imsettings-server.c
+++ b/imsettings-daemon/imsettings-server.c
@@ -1039,9 +1039,11 @@ imsettings_server_bus_method_call(GDBusConnection       *connection,
 	d(g_print("%s: sender[%s] path[%s] iface[%s] method[%s]\n", __PRETTY_FUNCTION__, sender, object_path, interface_name, method_name));
 
 	if (g_strcmp0(method_name, "StopService") == 0) {
+		value = g_variant_new_boolean(TRUE);
 		g_dbus_method_invocation_return_value(invocation,
-						      g_variant_new_boolean(TRUE));
+						      g_variant_new_tuple(&value, 1));
 		g_signal_emit(server, signals[SIG_DISCONNECTED], 0, NULL);
+		value = NULL;
 	} else if (g_strcmp0(method_name, "GetVersion") == 0) {
 		value = g_variant_new_uint32(IMSETTINGS_SETTINGS_API_VERSION);
 	} else if (g_strcmp0(method_name, "GetInfoVariants") == 0) {
-- 
2.17.1

